sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -650,15 +646,15 @@ object FoldablePropagation extends Rule[LogicalPlan] {
 
     if (foldableExprSet.isEmpty) {
       plan
-    } else if (hasCommandOrUnion(plan)) {
-      replaceInSubplan(plan)
     } else {
       val foldableMap = AttributeMap(foldableExprSet.toSeq.map(_.asInstanceOf[Alias])
         .map(a => (a.toAttribute, a.child)))
 
-      CleanupAliases(plan.transformAllExpressions {
-        case a : AttributeReference if foldableMap.contains(a) =>
-          Alias(foldableMap.getOrElse(a.toAttribute, a), a.name)()
+      CleanupAliases(plan.transformUp {
+        case p: LogicalPlan if !hasCommandOrUnion(p) => p.transformAllExpressions {
+          case a: AttributeReference if foldableMap.contains(a) =>
+            Alias(foldableMap.getOrElse(a.toAttribute, a), a.name)()
+        }
```

Comment:
As you told me, for the `stop` flag, I can change the above code more like the following. IMHO, the above code is simpler. The query optimization speed is more important? If then, I'll update that too.
```
var stop = false
CleanupAliases(plan.transformUp {
  case p: LogicalPlan if !stop && hasCommandOrUnion(p) => 
    stop = true
    p
  case p: LogicalPlan if !stop && !hasCommandOrUnion(p) => p.transformAllExpressions {
    case a: AttributeReference if foldableMap.contains(a) => 
      Alias(foldableMap.getOrElse(a.toAttribute, a), a.name)() 
  }
})
```